modules/internal/Bytes.chpl

```diff
@@ -0,0 +1,1631 @@
+  inline proc +(x: enumerated, s: _bytes) return concatHelp(x, s);
+  inline proc +(s: _bytes, x: bool) return concatHelp(s, x);
+  inline proc +(x: bool, s: _bytes) return concatHelp(x, s);
+
+  // Relational operators
+  private inline proc _strcmp_local(a, b) : int {
+    // Assumes a and b are on same locale and not empty.
+    const size = min(a.len, b.len);
+    const result =  c_memcmp(a.buff, b.buff, size);
+
+    if (result == 0) {
+      // Handle cases where one string is the beginning of the other
+      if (size < a.len) then return 1;
+      if (size < b.len) then return -1;
+    }
+    return result;
+  }
+
+  private inline proc _strcmp(a, b) where a.type==_bytes||a.type==string &&
+                                          b.type==_bytes||b.type==string : int {
+    if a.locale_id == chpl_nodeID && b.locale_id == chpl_nodeID {
+      // it's local
+      return _strcmp_local(a, b);
+    } else {
+      var localA: _bytes = a.localize();
+      var localB: _bytes = b.localize();
+      return _strcmp_local(localA, localB);
+    }
+  }
+
+  pragma "no doc"
+  proc ==(a: _bytes, b: _bytes) : bool {
+    // At the moment, this commented out section will not work correctly. If a
+    // and b are on the same locale, we will go to that locale, but an autoCopy
+    // will localize a and b, before they are placed into the on bundle,
+    // causing us to access garbage data inside the doEq routine. Always
+    // localize for now.
+    //
+    /* if a.locale_id == b.locale_id {
+      var ret: bool = false;
+      on __primitive("chpl_on_locale_num",
+                     chpl_buildLocaleID(a.locale_id, c_sublocid_any)) {
+        ret = doEq(a, b);
+      }
+      return ret;
+    } else { */
+
+    return _strcmp(a, b) == 0;
+  }
+
+  pragma "no doc"
+  proc ==(a: _bytes, b: string) : bool {
+    return _strcmp(a, b) == 0;
+  }
+
+  pragma "no doc"
+  proc ==(a: string, b: _bytes) : bool {
+    return _strcmp(a, b) == 0;
+  }
+
+  pragma "no doc"
+  inline proc !=(a: _bytes, b: _bytes) : bool {
+    return _strcmp(a, b) != 0;
+  }
+
+  pragma "no doc"
+  inline proc !=(a: _bytes, b: string) : bool {
+    return _strcmp(a, b) != 0;
+  }
+
+  pragma "no doc"
+  inline proc !=(a: string, b: _bytes) : bool {
+    return _strcmp(a, b) != 0;
+  }
+
+
+  // ASCII helpers
+
+  private param asciiSpace:byteType = 32;
+  private param asciiHTab:byteType = 9;
+  private param asciiVTab:byteType = 11;
+  private param asciiLineFeed:byteType = 10;
+  private param asciiFormFeed:byteType = 12;
+  private param asciiCarriageReturn:byteType = 13;
+  private param asciiDelete:byteType = 127;
+
+  private param asciiA = 65;
+  private param asciiZ = 90;
+
+  private param asciia = 97;
+  private param asciiz = 122;
+
+  private param ascii0 = 48;
+  private param ascii9 = 57;
+
+  private param asciiCaseDiff = asciia-asciiA;
+
+  private inline proc ascii_isWhitespace(c: byteType): bool {
```

Comment:
I would much rather see these functions use the C `isspace()` et al. to make sure we align with what people would normally expect, and remove the possibility of forgetting a case.  Our code should not need to know the ASCII value of a space, for example.

Reply:
@dmk42 - I think we are planning to make the Chapel programs work (only with) UTF-8 encoding for strings and to ignore the C LOCALE setting for strings specifically - although we might decide to use the C LOCALE settings for setting up I/O files/channels and do transcoding on I/O. If we were using the C LOCALE to do transcoding on I/O, we can't assume that the C LOCALE is configured for UTF-8 or ASCII or anything compatible, can we?

I don't personally view EBCDIC as an important use case for Chapel, but as a thought experiment, what would happen if the C LOCALE was EBCDIC? We could either:
 a. Fail to launch the program and report it as an unsupported configuration
 b. Support transcoding to/from Unicode in I/O of character data/strings

In scenario a. we could in fact use `isspace()` here. But in scenario b. we could not, because it would assume we were asking about something EBCDIC-encoded rather than ascii-encoded (which is what these methods on the `bytes` type assume).

(BTW is there an issue open asking if we will do a. or b. or something else? If not it's probably a good idea to make one).

Since I don't know if we are doing a. or b., I think Engin should not use `ctype.h` for these right now. If we decide on a., we could update this code to use ctype.h.

Reply:
FWIW, I had already implemented those changes before this comment but haven't pushed them. I can remove those commits easily from my local branch.

As for the particular topic, I am not quite familiar with the implications of either approach, however, using ctype would make the implementation closer to the string implementation and safer against possible subtle mistakes. But if we are sticking with the current implementation I can check closer for those subtleties.

Reply:
@e-kayrakli - arguably if we want to support the scenario b. in my comment, we will need to replace the similar calls in string with new calls that always operate with Unicode characters, and then you could use those as well (because Unicode contains ASCII). So either way could be OK for the near-term.

Reply:
@mppf -- that's true. So, I am leaning towards using `ctype` currently, just for increased confidence in the implementation and similar implementation to current string type.

Reply:
Currently we use `wctype` for the wide-character versions of these predicates, which means we already must have the character locale set to UTF-8, where ASCII is the low seven bits, so `ctype` fits into the current usage.